graph/src/components/metrics/mod.rs

```diff
@@ -11,57 +11,203 @@ pub mod stopwatch;
+        )?);
+        self.register(name, counters.clone());
+        Ok(counters)
+    }
 
-    fn new_histogram(
+    fn new_subgraph_counter_vec(
         &self,
-        name: String,
-        help: String,
-        const_labels: HashMap<String, String>,
+        name: &str,
+        help: &str,
+        subgraph: &str,
+        variable_labels: Vec<String>,
+    ) -> Result<Box<CounterVec>, PrometheusError> {
+        let opts = Opts::new(name.clone(), help).const_labels(subgraph_labels(subgraph));
+        let counters = Box::new(CounterVec::new(
+            opts,
+            variable_labels
+                .iter()
+                .map(|s| s.as_str())
+                .collect::<Vec<&str>>()
+                .as_slice(),
+        )?);
+        self.register(name, counters.clone());
+        Ok(counters)
+    }
+
+    fn new_subgraph_histogram(
+        &self,
+        name: &str,
+        help: &str,
+        subgraph: &str,
         buckets: Vec<f64>,
-    ) -> Result<Box<Histogram>, PrometheusError>;
+    ) -> Result<Box<Histogram>, PrometheusError> {
+        let opts = HistogramOpts::new(name.clone(), help)
+            .const_labels(subgraph_labels(subgraph))
+            .buckets(buckets);
+        let histogram = Box::new(Histogram::with_opts(opts)?);
+        self.register(name, histogram.clone());
+        Ok(histogram)
+    }
 
     fn new_histogram_vec(
         &self,
-        name: String,
-        help: String,
-        const_labels: HashMap<String, String>,
+        name: &str,
+        help: &str,
         variable_labels: Vec<String>,
         buckets: Vec<f64>,
-    ) -> Result<Box<HistogramVec>, PrometheusError>;
+    ) -> Result<Box<HistogramVec>, PrometheusError> {
+        let opts = Opts::new(name.clone(), help);
+        let histograms = Box::new(HistogramVec::new(
+            HistogramOpts {
+                common_opts: opts,
+                buckets,
+            },
+            variable_labels
+                .iter()
+                .map(|s| s.as_str())
+                .collect::<Vec<&str>>()
+                .as_slice(),
+        )?);
+        self.register(name, histograms.clone());
+        Ok(histograms)
+    }
 
-    fn unregister(&self, metric: Box<dyn Collector>);
+    fn new_subgraph_histogram_vec(
+        &self,
+        name: &str,
+        help: &str,
+        subgraph: &str,
+        variable_labels: Vec<String>,
+        buckets: Vec<f64>,
+    ) -> Result<Box<HistogramVec>, PrometheusError> {
+        let opts = Opts::new(name.clone(), help).const_labels(subgraph_labels(subgraph));
+        let histograms = Box::new(HistogramVec::new(
+            HistogramOpts {
+                common_opts: opts,
+                buckets,
+            },
+            variable_labels
+                .iter()
+                .map(|s| s.as_str())
+                .collect::<Vec<&str>>()
+                .as_slice(),
+        )?);
+        self.register(name, histograms.clone());
```

Comment:
Have you tested this with multiple subgraphs? I'm wondering if registering the same name twice won't fail.

Reply:
I have now run it in the integration cluster, and that worked without errors. I'll let it run over the weekend, but it looks like (after the additions I made) it's all working.